core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
I am no sure that such semantic is correct for me. I will explain design principles I followed and ask you to correct me if these principles was wrong.

As an API designer I am sure that graph stages should have their lifecycle rules and be composable in terms of lifecycle. It means for me that the user should have guaranties about how stages will work and handle their resources. Since very first version (0.3) I was sure that such contract looks like this:
- stage should clean up resources on completion
- indication of completion is onComplete or onFailure on all outputs

This contract is very clear and simple and I always follow it. This allows for example easily have endless `Source[Source[Source[T]]]` and be sure about resources usage, because all I have to do is to control number of simultaneous working instances of `Source`. Option 1 was designed in such way - you can easy be notified when resources released. 

With timeout stop semantic is also possible to be notified about stage clean (with `Future` from `stop` for example), but for me it is not so natural as notification via stage completion.

Another possible problem is having _rich_ (I will messages rich if it is not plain data) message which functionality depends on something that is not visible from the current context. Using such kind of messages may introduce class of problems that can not be found at compile time. You may end up with a message that have no chance to be completed and you have no way to be notified that your client is closed.